### PR TITLE
Fix: honour .impeccable detector ignores in the live overlay (#639)

### LINKED ISSUE #639: Live overlay ignores .impeccable/config.json: requestDetectScan never sends disabledRules
**Version:** skill 4.1.1

## Summary

The live overlay's detect scan runs **unfiltered**. `detector.ignoreRules` in `.impeccable/config.json` applies to `impeccable detect` and to the edit hook, but never to the in-browser overlay, so a project that has deliberately waived a rule still sees it on every page, forever.

The suppression mechanism already exists and is already honoured by the detector. It is simply never populated on this path.

## Where it breaks

`skill/scripts/live-browser.js`, `requestDetectScan()`:

```js
window.postMessage({
  source: 'impeccable-command',
  action: 'scan',
  config: { scanId },          // ← only the scan id
}, '*');
```

`cli/engine/browser/injected/index.mjs`, `collectBrowserFindings()`:

```js
const _disabled = EXTENSION_MODE ? (window.__IMPECCABLE_CONFIG__?.disabledRules || []) : [];
const _ruleOk = (id) => !_disabled.length || !_disabled.includes(id);
```

The overlay loads `detect.js` with `dataset.impeccableExtension = 'true'`, so `EXTENSION_MODE` is true and the branch is live — but `config.disabledRules` is never sent, so `_disabled` is always `[]` and every rule fires.

`skill/scripts/live-server.mjs` serves `/detect.js` verbatim and builds `/live.js` with a prelude of `window.__IMPECCABLE_*` globals, so there is a natural place to hand the project config across.

**Precedent:** `extension/devtools/panel.js` already drives this exact field (`disabledRules`) from its own checkbox UI. The field works; live mode just doesn't feed it.

## Reproduction

1. In a project, `node .claude/skills/impeccable/scripts/hook-admin.mjs ignore-rule <some-rule>`
2. Confirm `impeccable detect <file>` no longer reports it.
3. Start live mode, open the page, run the overlay's detect scan.
4. The rule still reports.

Measured on a 28-page static site with two rules waived (`ai-color-palette`, `gradient-text`): homepage showed **46** findings in the overlay against **13** from the same detector with `disabledRules` passed. Site-wide, 1,046 of 1,792 findings were ones the project had already dispositioned.

## Effect

The overlay is the surface a designer actually looks at while working, and it is the one place waivers do not reach. On a project with a brand colour the palette rules flag — ours is a logo teal at hue 173, inside the rule's 160–200 "cyan" band, so `a.logo` and `img.logo-full` are themselves reported — the noise floor makes the panel unusable for spotting anything new.

## Suggested fix

Thread the project config into the prelude `assembleLiveBrowserScript` already builds, and pass it through:

```js
config: { scanId, disabledRules: window.__IMPECCABLE_IGNORE_RULES__ || [] },
```

Note `tests/live-browser-regression.test.mjs:717` asserts the current shape exactly (`config: \{ scanId \}`), so that assertion needs relaxing as part of any fix.

## Open question, and the reason I am filing rather than sending a patch

`disabledRules` is rule-level and all-or-nothing. Project config also carries **`detector.ignoreValues`**, which is file- and value-scoped (`{rule, value, files[]}`), and that has no browser equivalent. Supporting it in the overlay means either extending the detector's browser config contract or filtering results client-side in `live-browser.js`. There is also `config.local.json`, the gitignored per-developer override, which the docs say wins.

That boundary feels like a maintainer's call rather than something to assume in a PR. Happy to send one for whichever shape you prefer — including the narrow rules-only version, which is a couple of lines and already fixes the majority of the noise.

### PR DESCRIPTION
**Pull request by Guitaraholic**
Fixes #639. This is the patch from the issue's follow-up comment, restructured to fit the repo's conventions rather than pasted in as posted. Shape deltas from that diff are listed at the end so review can focus on them.

## What it does

The live overlay's detect scan ran unfiltered: `requestDetectScan()` posted only `{ scanId }`, so `detector.ignoreRules` and `detector.ignoreValues` reached `impeccable detect` and the edit hook but never the overlay. Now:

- **`skill/scripts/live-server.mjs`** serializes the project's detector waivers into the `/live.js` prelude as `window.__IMPECCABLE_PROJECT_IGNORES__`, read per request so a config edit lands on the next tab reload. It reuses `readConfig` from `hook-lib.mjs`, so the `config.json` + `config.local.json` merge and its type guards are the hook's own, not a second implementation. The served-root prefixes come from the inject config's `files` globs via `resolveLiveConfigPath`.
- **`skill/scripts/live-browser-ignores.js`** (new script part) resolves that config against the page URL when a scan starts. It deliberately mirrors `cli/lib/impeccable-config.mjs`: the same glob dialect as `globToRegex` (including `{a,b}`), the same path-suffix matching as `findingMatchesScopedIgnoreFile`, and the same refusal to apply an unscoped `value: "*"` entry. Being a separate part makes the trickiest logic unit-testable in Node.
- **`skill/scripts/live-browser.js`**'s `requestDetectScan()` calls the resolver and sends `disabledRules` + `disabledValues` with the scan, guarded so a stale cached `live.js` still scans, just unfiltered as before.
- **`cli/engine/browser/injected/index.mjs`** applies `disabledValues` where the findings are assembled (the overlay draws its own markers from `collectBrowserFindings`, so a rule-level filter alone can never reach a value-scoped waiver). Bundle regenerated with `build:browser`.

The roots prefix is the part worth reading twice, per the issue: it must come from the inject config's own `files` globs, never be derived from the ignore globs. Deriving it from the ignore globs lets one entry scoped to `prototype/library/**` lend its prefix to every page and suppress its rule site-wide, and it looks like success because the finding counts go down.

## Tests

- `tests/live-browser-ignores.test.mjs` (new): the resolver's edge cases from the issue: missing/stale globals, `ignoreRules` as a string, malformed entries, `roots` as a number, root and directory URLs, malformed percent-escapes, glob metacharacters in filenames, suffix matching, alternation, and the roots trap above.
- `tests/detect-antipatterns-browser.test.mjs`: new extension-mode test that `disabledValues` suppress exactly the waived findings and unrelated findings survive.
- `tests/live-browser-regression.test.mjs:717` now pins the new scan config shape (the issue flagged this assertion), plus a pin that the resolver is guarded for stale caches.
- `tests/live-browser-script-parts.test.mjs`: new part and prelude assertions.

Suites run locally, all green: `test:live` (859), `test:detector` (552), `test:core` (1160), `test:live-e2e` (35 pass, 1 env skip), `bun run build`, and `build:browser` re-run to confirm the committed bundle is stable. Also verified against the 28-page site from the issue: a harness comparing the browser resolver against `filterDetectionFindings` across every entry and page of that project's real config (185 aggregate decisions) found zero disagreements.

## Deltas from the diff posted in the issue

- Resolver extracted into a script part instead of inlined in `requestDetectScan()`; server config read through `hook-lib.readConfig` instead of hand-rolled JSON reads; prelude emission moved into `assembleLiveBrowserScript` where the other globals live.
- Three parity fixes found while writing the unit tests: the glob matcher now supports `{a,b}` and matches path suffixes exactly as the CLI does, and an unscoped `value: "*"` entry is now skipped (the CLI's `isIgnoredFindingValue` never applies one; the posted diff wrongly treated it as a project-wide rule disable).
- `createdAt`/`reason` fields are stripped before serializing entries into the page.

## Known gaps, kept out of scope deliberately

- `bounce-easing`: the motion value extractor is not mirrored into the browser bundle, so a value-scoped bounce-easing waiver only matches when the finding carries `ignoreValue` directly. Happy to port `extractMotionIgnoreValue` in a follow-up if you'd rather share that code.
- `design-system-color` matches on the normalized string without the CLI's color-equality fallback (a waiver written as `rgb()` will not match a finding reported as hex).
- Entries scoped to asset files (e.g. a `.css` file) resolve against the page URL here, so they apply only in the CLI. Page-scoped resolution can't see which stylesheet a computed style came from.

If you'd rather keep the browser contract rule-level and solve `ignoreValues` another way, the value-level half (`disabledValues` in the detector plus its tests) lifts out cleanly.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes which detector findings the overlay hides. Incorrect glob/root matching could suppress real issues or fail to honour project waivers, but this is not auth or data-handling.
> 
> **Overview**
> Live overlay detect scans now honour project detector waivers from `.impeccable/config.json` (and `config.local.json`), matching CLI/edit-hook filtering (issue #639).
> 
> The live server serializes `ignoreRules`/`ignoreValues` plus served-root prefixes into `/live.js`. A new `live-browser-ignores` script part resolves them against the current URL (CLI glob dialect, path-suffix matching, no unscoped `value: "*"`). `requestDetectScan()` sends `disabledRules` and `disabledValues`; the browser detector then drops matching findings. Roots come from inject `files` globs, not ignore globs, so a scoped waiver cannot leak site-wide.
> 
> Known gaps left out of scope: bounce-easing motion extraction, design-system-color rgb/hex equality, and CSS-file-scoped entries (page URL only).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 092561f1c24f7c68658347481542d8fafd57148e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by Guitaraholic** [comment]
The bot's finding is real, and 68d62bb fixes it.

**What was wrong.** If the live config lists pages under more than one folder (say `src/**/*.html` and `public/**/*.html`), the overlay treated a URL like `/foo.html` as `src/foo.html` and `public/foo.html` at the same time. So a waiver written only for `src/foo.html` could hide a finding on the page actually served from `public/foo.html`. That fails in the worst direction: a real finding disappears and nothing says so.

**Why the overlay can't just look up the right file.** The live server does not serve the pages. The project's own dev or static server does, and its URL-to-file mapping is invisible from here. So when more than one folder could hold a page, the overlay can never be certain which file it is looking at.

**What the fix does.** It stops guessing. A file-scoped waiver now applies only when one of two things is true:

- it matches the URL path itself, which holds no matter which folder serves the page, or
- it matches under every configured folder, so no possible reading disagrees.

Anything ambiguous shows the finding instead of hiding it, which is also what the CLI reports for the file really being served. With a single configured root, the common case, the new rule reduces to the old behaviour exactly, and the tests prove it. Projects that really serve from several folders keep three ways to write a waiver that still applies: name the file under each folder, use the bare path (`foo.html`), or use `**/foo.html`.

Two new unit tests pin the ambiguous case and the safe spellings. The live suite and the parity harness from the PR description still pass.

One related gap I noticed while checking this, left out on purpose: `detector.ignoreFiles` (skip a whole file) still has no overlay equivalent, so a page the CLI skips entirely is still scanned live. Honouring it would make an ignored page report "no issues", which reads as clean rather than skipped, so the presentation feels like your call. Happy to take that as a follow-up.

**Comment by Guitaraholic** [comment]
Both findings are right, and 092561f fixes them.

**The high one.** The previous commit over-corrected. When one tree is listed at two depths (`prototype/*.html` plus `prototype/library/**/*.html`), the resolver derived two roots and asked a waiver to match under both. A normal waiver like `prototype/index.html` can't do that, so it stopped applying anywhere. Confirmed with a repro before fixing.

**The rule both reviews were circling is simpler.** One live session is served by one server, so a single document root must sit at or above every configured page. That means the only prefix the resolver can safely assert is the deepest common ancestor of the glob roots:

- Nested roots collapse into their shared tree, so normal project-relative waivers keep applying.
- Disjoint roots (`src/` and `public/`) share nothing, so no prefix is asserted and only the URL path itself matches. The earlier fix stays intact: a `src/foo.html` waiver still cannot hide a finding on a page served from `public/foo.html`.

This also let me delete the match-under-every-root logic from the last commit; with one asserted prefix, plain matching is enough, so the resolver is now simpler than it was before either review round.

The low one is fixed too: the new test file derives the repo root from `import.meta.url` instead of `process.cwd()`.

New tests pin both scenarios (nested roots and disjoint roots), and the live suite plus the parity harness from the PR description still pass.

**Comment by Guitaraholic** [comment]
Anything else needed to get this merged?

**Comment by abdulwahabone** [comment]
> Anything else needed to get this merged?

Let me test this shortly

**Comment by abdulwahabone** [comment]
Tested locally, seems to be working well. Nice work @Guitaraholic 

image

File: skill/scripts/live-server.mjs [hunk 1]
```diff
@@ -686,6 +688,51 @@ function isLoopbackOrigin(origin) {
...
+    const live = JSON.parse(fs.readFileSync(configPath, 'utf-8'));
+    const files = Array.isArray(live?.files) ? live.files : [];
+    return [...new Set(files
+      .filter((glob) => typeof glob === 'string' && glob)
+      .map((glob) => {
+        const wildcardAt = glob.search(/[*?{]/);
+        const head = wildcardAt === -1 ? glob : glob.slice(0, wildcardAt);
+        const cut = head.lastIndexOf('/');
+        return cut > -1 ? head.slice(0, cut + 1) : '';
+      }))];
```
[the remark this answers is not in the document]
Comment (abdulwahabone) [reply skill/scripts/live-server.mjs:729]: Lets resolve it for now, I will harden it on the next PR
